multiplexer/main.go

```diff
@@ -35,44 +61,71 @@ func getContent(client *http.Client, url string) ([]byte, error) {
 
 func main() {
 	var urls URLs
-	var wg sync.WaitGroup
-	var mu sync.RWMutex
+	var jwg sync.WaitGroup
+	var rwg sync.WaitGroup
+	jobs := make(chan Job)
+	result := make(chan Response)
+	workers := 5
+
 	done := make(chan os.Signal, 1)
 	signal.Notify(done, os.Interrupt, syscall.SIGINT, syscall.SIGTERM)
 
-	client := http.Client{
-		Timeout: time.Second,
-	}
+	client := newClient()
 
 	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
 		if err := json.NewDecoder(r.Body).Decode(&urls); err != nil {
 			return
 		}
 
 		if len(urls.List) > 20 {
+			w.WriteHeader(http.StatusInternalServerError)
 			w.Write([]byte("amount of url's exceeded limit"))
 			return
 		}
 
-		for _, url := range urls.List {
-			wg.Add(1)
-			go func(u string) {
-				mu.RLock()
-				content, err := getContent(&client, u)
-				mu.RUnlock()
-				
-				if err != nil {
-					w.Write([]byte(err.Error()))
+		//reciving result, sorting, and then sending to client
+		rwg.Add(1)
+		go func() {
+			responses := make([]string, len(urls.List))
+			for r := range result {
+				responses[r.index] = string(r.content)
+			}
+
+			for i, content := range responses {
+				w.Write([]byte(content))
+				fmt.Printf("wrote content of %s\n", urls.List[i])
+			}
+			rwg.Done()
+		}()
+
+		//pulling from channel and working, and then sending to be sorted
+		jwg.Add(workers)
+		for i := 0; i < workers; i++ {
+			go func() {
+				for url := range jobs {
+					content, err := client.getContent(url.site)
+					if err != nil {
+						w.WriteHeader(http.StatusInternalServerError)
+						w.Write([]byte(err.Error()))
+						return
+					}
+					result <- Response{index: url.index, content: content}
 				}
-
-				mu.Lock()
-				w.Write(content)
-				mu.Unlock()
-
-				wg.Done()
-			}(url)
+				jwg.Done()
+			}()
 		}
-		wg.Wait()
+
+		//sending the url's to the channel to start aggregating
+		go func() {
+			for i, url := range urls.List {
+				jobs <- Job{index: i, site: url}
+			}
+			close(jobs)
+		}()
+
+		jwg.Wait()
+		close(result)
```

Comment:
not sure when to close result channel. i get panic error if i try to send another post request

Reply:
Please look above comment